### ISSUE TITLE
Offload without Checkpointing: No Reduction in Max GPU Memory

### ISSUE THREAD

**Issue by awgu** [facebookresearch/fairscale#826]
**Overview**
I am exploring some techniques for saving GPU memory during training, including [Fairscale's `OffloadModel`](https://github.com/facebookresearch/fairscale/blob/main/fairscale/experimental/nn/offload.py). As an initial exploration, I have been testing with a ResNet152 implementation based on [this repo](https://github.com/Lyken17/pytorch-memonger/blob/master/memonger/resnet.py) (instead of through `torchvision`) since the repo's implementation uses `nn.Sequential` for each of the four blocks, meaning it is amenable to `OffloadModel`.

Specifically, I am wrapping each `nn.Sequential` instance with `OffloadModel`, as seen [here](https://github.com/andwgu/offload/blob/88ecc59cdb82d712d7a52b01dcca3308cff9c15f/resnet.py#L125). However, I am not seeing any meaningful reduction in max GPU memory usage compared to the baseline model, even when varying the `num_slices` argument. (Though, setting `checkpoint_activation=True` does reduce the GPU memory usage as expected.)

Any help in determining why this may be is much appreciated. 

**Details**
Concretely, I am comparing (1) a naive layer-to-layer (L2L) offloading, where only a *single* layer is moved to GPU memory at a time, (2) gradient/activation checkpointing, and (3) Fairscale's `OffloadModel` without checkpointing, and (4) Fairscale's `OffloadModel` with checkpointing.

For ResNet152 with a batch size of 32, the max GPU memory usage is as follows:

| Method | Max GPU Memory Usage (MB) |
|:---:|:---:|
| Baseline | 3105.648 |
| L2L | 1054.011 |
| Checkpoint | 904.423 |
| Offload w/o Checkpoint | 2931.984 |
| Offload w/ Checkpoint | 725.737 |

These can be reproduced by running `python main.py <...>` where `<...>` is either `baseline`, `l2l`, `checkpoint`, `offload`, or `offload_checkpoint` from [my repository](https://github.com/andwgu/offload) (which assumes PyTorch v1.9, `torchinfo`, and access to a single GPU).

The only change I made to Fairscale's `offload.py` is [adding an `is_leaf()` check in `OffloadFunction.backward()`](https://github.com/andwgu/offload/blob/88ecc59cdb82d712d7a52b01dcca3308cff9c15f/offload.py#L230-L232) to avoid erroring when setting `requires_grad`:
```
for i, need_grad in enumerate(ctx.grad_requirements):
    if inputs[i].is_leaf:
        inputs[i].requires_grad = need_grad
```

Intuitively, we would expect to see "Offload w/o Checkpoint" to be closer to 1000 MB rather than be almost the same as "Baseline". The relative results for "L2L", "Checkpoint", and "Offload w/ Checkpoint" seem to sanity check since "L2L" is the extreme version of "Offload w/o Checkpoint" and additionally checkpointing should further reduce the max GPU memory usage -- hence, "Offload w/ Checkpoint" achieves the lowest max GPU memory usage. However, it almost seems that the offloading is not being done correctly when `checkpoint_activation=False`.

Please feel free to let me know if I should provide any other information, and thanks for [releasing `OffloadModel`](https://github.com/facebookresearch/fairscale/issues/337#issuecomment-945751906).

**Comment by nyngwang** [comment]
So why this is closed?
